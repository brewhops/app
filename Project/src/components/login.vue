<template>
<div>
  <div class="header">
    <h2>Ninkasi Brew Hops</h2>
  </div>
  <div class="column-1">
    <div id="login">
      <img src="../assets/logo.png" class="logo-med">
      <h2>Login</h2>
      <input v-model.lazy="username" placeholder="username">
      <span>{{ feedback.username }}</span>
      <input v-model.lazy="password" placeholder="password" type="password">
      <span>{{ feedback.password }}</span>
      <button v-on:click="submit">Submit</button>
      <button v-if="isAdmin" v-on:click="validateAdmin" type="button" name="button">Admin Page</button>
    </div>
  </div>
</div>
</template>

<script lang="ts">

import router from "../router/index.js"
import CryptoJS from "crypto-js"
import Cookie from "js-cookie"

interface IData {
  mobile: any;
  isAdmin: any;
  username: any;
  password: any;
  encryptedPassword: any;
  feedback: any;
  submitLink: any;
}

interface ILogin {
  name: any;
  data: () => IData;
  watch: any;
  beforeMount: any;
  methods: any;
  mobile?: any;
  encryptedPassword?: any;
  feedback?: any;
  submitLink?: any;
  sendToHome?: any;
  $http?: any;
  employees?: any;
}

const login: ILogin = {
  name: 'login',
  data: function() {
    return {
      employees: {},
      mobile: false,
      isAdmin: false,
      username: '',
      password: '',
      encryptedPassword: '',
      feedback: {
        username: '',
        password: '',
      },
      submitLink: '',
    };
  },
  watch: {
    username: function() {
      // shorten our username variable for readability
      const username = this.username;


      // if the field is empty, clear the feedback
      // if the username is less than 3 or greater than 20, error
      // if the username is only numbers and letters
      // else there are illegal characters
      if (username.length === 0) {
        this.feedback.username = ''
      } else if (username.length < 3 || username.length > 20) {
        this.feedback.username = 'username must be between 3 and 20 characters'
      } else if (username.match('^[0-9A-z]+$')) {
        this.feedback.username = ''
      } else {
        this.feedback.username = 'username must only be digits and letters'
      }

      // if the username is admin
      for (var x in this.employees) {
        if (this.employees[x].username === username) {
          this.encryptedPassword = this.employees[x].password
          if (this.employees[x].admin === true) {
            this.isAdmin = true
          }
        }
      }
    },
    password: function() {
      // shorten our username variable for readability
      const password = this.password

      // if the field is empty, clear the feedback
      // if the password is less than 6 or greater than 25, error
      // if the password is numbers, letters and some select special characters
      // else there are illegal characters
      if (password.length === 0) {
        this.feedback.password = ''
      } else if (password.length < 6 || password.length > 25) {
        this.feedback.password = 'password must be between 6 and 25 characters'
      } else if (password.match('^[0-9A-z@#$%*^!\-=&]+$')) {
        this.feedback.password = ''
      } else {
        this.feedback.password = 'password must only be digits and letters'
      }
    }
  },
  beforeMount() {
    if (/iPhone|iPad|iPod|Android|webOS|BlackBerry|BB|PlayBook|IEMobile|Windows Phone|Kindle|Silk|Opera Mini/i.test(navigator.userAgent)) {
      this.mobile = true
      this.submitLink = '/home-mobile'
    } else {
      this.submitLink = '/home'
    }

    // if the cookie has login information in it already
    // then send us straight to the home page
    if (Cookie.get('loggedIn')) {
      this.sendToHome()
    }

    //get users from heroku
    this.$http.get(process.env.API + '/employees').then(response => {
      // get body data
      this.employees = response.body;
    }, response => {
      console.log('Response error, cant access employees page', response);
    });
  },
  methods: {
      submit: function () {
        // for each employee in the database
        for(var x in this.employees) {
          // if the username matches the inputed username
          if(this.employees[x].username === this.username) {
            if (this.validPassword(this.username, this.encryptedPassword, this.password)) {
              this.createCookie(this.username, this.isAdmin);
              // send us over to the home page
              this.sendToHome()
            }
          }
        }
        // we were never redirected, so the login was invalid
        this.feedback.password = 'Invalid Login'
      },
      validateAdmin: function () {
        // saving for reference: to encrypt
        // CryptoJS.AES.encrypt(password, username).toString();

        if (this.validPassword(this.username, this.encryptedPassword, this.password)) {
          this.createCookie(this.username, this.isAdmin);
          router.push("admin")
        }
      },
      // redirect over to the home page
      sendToHome: function() {
        if (this.mobile) {
          router.push("home-mobile")
        } else {
          router.push("home")
        }
      },
      // decrypt the password and check if it is valid
      validPassword: function(username, encryptedPassword, enteredPassword) {
        // decrypt the user password from the database
        var decryptedPassword = CryptoJS.AES.decrypt(
            encryptedPassword,
            username
          ).toString(CryptoJS.enc.Utf8)

        // if the decrypted password at that same point matches the user password
        if(decryptedPassword === enteredPassword) {
          return true
        }
      },
      createCookie: function(username, adminStatus) {
        if (!Cookie.get('loggedIn')) {
          Cookie.set('loggedIn', {
            'username': username,
            'admin': adminStatus,
          })
        }
      }
    }
};

export default login;
</script>

<style lang="stylus" scoped>
#login
  display flex
  flex-direction column
  align-items center
  width 30%
  min-width 300px
  img
    margin 20px
  h2
    padding 0
    margin 10px

</style>
