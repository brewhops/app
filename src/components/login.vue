<template>
  <div>
    <div class="header">
      <h2>Ninkasi Brew Hops</h2>
    </div>
    <div class="column-1">
      <div id="login">
        <img src="../assets/logo.png" class="logo-med" />
        <h2>Login</h2>
        <input v-model.lazy="username" v-on:keyup.enter="submit" placeholder="username" />
        <span>{{ feedback.username }}</span>
        <input
          v-model.lazy="password"
          v-on:keyup.enter="submit"
          placeholder="password"
          type="password"
        />
        <span>{{ feedback.password }}</span>
        <button v-on:click="submit">Submit</button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import router from '../router/index.js';
import CryptoJS from 'crypto-js';
import Cookie from 'js-cookie';

interface ILoginState {
  mobile: boolean;
  isAdmin: boolean;
  username: string;
  password: string;
  feedback: {
    username: string;
    password: string;
  };
  submitLink: string;
}

export default Vue.extend({
  name: 'login',
  data(): ILoginState {
    return {
      mobile: false,
      isAdmin: false,
      username: '',
      password: '',
      feedback: {
        username: '',
        password: ''
      },
      submitLink: ''
    };
  },
  watch: {
    async username(): Promise<void> {
      // shorten our username variable for readability
      const validUsername = false;
      const { username } = this;

      // if the field is empty, clear the feedback
      // if the username is less than 3 or greater than 20, error
      // if the username is only numbers and letters
      // else there are illegal characters
      if (username.length === 0) {
        this.feedback.username = '';
      } else if (username.length < 3 || username.length > 20) {
        this.feedback.username = 'username must be between 3 and 20 characters';
      } else if (!username.match('^[0-9A-z]+$')) {
        this.feedback.username = 'username must only be digits and letters';
      } else {
        try {
          const response = await this.$http.get(`${process.env.API}/employees/admin/${username}`);
          this.isAdmin = response.data;
        } catch (err) {
          // tslint:disable-next-line:no-console
          console.error(err);
        }
      }
    },
    password(): void {
      // shorten our username variable for readability
      const { password } = this;

      // if the field is empty, clear the feedback
      // if the password is less than 6 or greater than 25, error
      // if the password is numbers, letters and some select special characters
      // else there are illegal characters
      if (password.length === 0) {
        this.feedback.password = '';
      } else if (password.length < 6 || password.length > 25) {
        this.feedback.password = 'password must be between 6 and 25 characters';
      } else if (password.match('^[0-9A-z@#$%*^!-=&]+$')) {
        this.feedback.password = '';
      } else {
        this.feedback.password = 'password must only be digits and letters';
      }
    }
  },
  beforeMount(): void {
    if (
      /iPhone|iPad|iPod|Android|webOS|BlackBerry|BB|PlayBook|IEMobile|Windows Phone|Kindle|Silk|Opera Mini/i.test(
        navigator.userAgent
      )
    ) {
      this.mobile = true;
      this.submitLink = '/home-mobile';
    } else {
      this.submitLink = '/home';
    }

    // if the cookie has login information in it already
    // then send us straight to the home page
    if (Cookie.getJSON('loggedIn')) {
      this.sendToHome();
    }
  },
  methods: {
    async submit(): Promise<void> {
      const { username, password: pw } = this;

      const password = CryptoJS.SHA3(pw).toString();

      try {
        const response = await this.$http.post(`${process.env.API}/employees/login/`, {
          username,
          password
        });
        const { id, token } = response.data;
        this.createCookie(id, this.username, this.isAdmin, token);
        this.sendToHome();
      } catch (err) {
        this.feedback.password = 'Invalid Login';
      }
    },
    // redirect over to the home page
    sendToHome(): void {
      if (this.mobile) {
        router.push('home-mobile');
      } else {
        router.push('home');
      }
    },
    createCookie(id, username, admin, token): void {
      if (!Cookie.getJSON('loggedIn')) {
        // tslint:disable-next-line:no-backbone-get-set-outside-model
        Cookie.set('loggedIn', {
          id,
          username,
          admin,
          token
        });
      }
    }
  }
});
</script>

<style lang="stylus" scoped>
#login {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 30%;
  min-width: 300px;

  img {
    margin: 20px;
  }

  h2 {
    padding: 0;
    margin: 10px;
  }
}
</style>
