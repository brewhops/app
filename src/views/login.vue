<template>
  <div>
    <div class="column-1">
      <div id="login">
        <img src="../assets/logo.png" class="logo-med" />
        <h2>Login</h2>
        <input v-model.lazy="username" v-on:keyup.enter="submit" placeholder="username" />
        <span>{{ feedback.username }}</span>
        <input
          v-model.lazy="password"
          v-on:keyup.enter="submit"
          placeholder="password"
          type="password"
        />
        <span>{{ feedback.password }}</span>
        <button v-on:click="submit">Submit</button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import Vue from 'vue';
import router from '@/router';
import CryptoJS from 'crypto-js';
import Cookie from 'js-cookie';
import { BrewhopsCookie } from '../types/index';

interface ILoginState {
  isAdmin: boolean;
  username: string;
  password: string;
  feedback: {
    username: string;
    password: string;
  };
  submitLink: string;
}

export default Vue.extend({
  name: 'login',
  data(): ILoginState {
    return {
      isAdmin: false,
      username: '',
      password: '',
      feedback: {
        username: '',
        password: ''
      },
      submitLink: ''
    };
  },
  watch: {
    async username(): Promise<void> {
      // shorten our username variable for readability
      const validUsername = false;
      const { username } = this;

      // if the field is empty, clear the feedback
      // if the username is less than 3 or greater than 20, error
      // if the username is only numbers and letters
      // else there are illegal characters
      if (username.length === 0) {
        this.feedback.username = '';
      } else if (username.length < 3 || username.length > 20) {
        this.feedback.username = 'username must be between 3 and 20 characters';
      } else if (!username.match('^[0-9A-z]+$')) {
        this.feedback.username = 'username must only be digits and letters';
      } else {
        try {
          const response = await this.$http.get(
            `${process.env.VUE_APP_API}/employees/admin/${username}/`
          );
          this.isAdmin = response.data;
        } catch (err) {
          // tslint:disable-next-line:no-console
          console.error(err);
        }
      }
    },
    password(): void {
      // shorten our username variable for readability
      const { password } = this;

      // if the field is empty, clear the feedback
      // if the password is less than 6 or greater than 25, error
      // if the password is numbers, letters and some select special characters
      // else there are illegal characters
      if (password.length === 0) {
        this.feedback.password = '';
      } else if (password.length < 6 || password.length > 25) {
        this.feedback.password = 'password must be between 6 and 25 characters';
      } else if (password.match('^[0-9A-z@#$%*^!-=&]+$')) {
        this.feedback.password = '';
      } else {
        this.feedback.password = 'password must only be digits and letters';
      }
    }
  },
  beforeMount(): void {
    this.submitLink = '/home';
    if (Cookie.getJSON('loggedIn')) {
      this.sendToHome();
    }
  },
  methods: {
    async submit(): Promise<void> {
      const { username, password: pw } = this;
      const password = CryptoJS.SHA3(pw).toString();

      try {
        const response = await this.$http.post(`${process.env.VUE_APP_API}/employees/login/`, {
          username,
          password
        });
        const { id, token, client_id } = response.data;
        this.createCookie({ id, username: this.username, admin: this.isAdmin, token, client_id });
        this.sendToHome();
      } catch (err) {
        this.feedback.password = 'Invalid Login';
      }
    },
    sendToHome(): void {
      router.push('home');
    },
    createCookie(cookie: BrewhopsCookie): void {
      if (!Cookie.getJSON('loggedIn')) {
        // TODO: create configurable expiration
        Cookie.set('loggedIn', { ...cookie }, { expires: 7 });
      }
    }
  }
});
</script>

<style lang="stylus" scoped>
#login {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 30%;
  min-width: 300px;

  img {
    margin: 20px;
  }

  h2 {
    padding: 0;
    margin: 10px;
  }
}
</style>
