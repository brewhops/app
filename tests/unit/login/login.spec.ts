import { shallowMount } from '@vue/test-utils'
import login from '@/views/login.vue'

describe('login.vue', () => {
  it('renders login page', () => {
    const wrapper = shallowMount(login);
    expect(wrapper).toMatchSnapshot();
  })
})
