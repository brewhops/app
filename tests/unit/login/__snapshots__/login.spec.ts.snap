// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login.vue renders login page 1`] = `
<div>
  <div class="column-1">
    <div id="login"><img src="../assets/logo.png" class="logo-med">
      <h2>Login</h2> <input placeholder="username"> <span></span> <input placeholder="password" type="password"> <span></span> <button>Submit</button>
    </div>
  </div>
</div>
`;
